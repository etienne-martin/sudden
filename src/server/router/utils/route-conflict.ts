import { CompiledEndpoints, convertFileNameToRoute } from "./index";

interface RegisteredRoutes {
  routeName: string;
  method: string;
  fileName: string;
}

// TODO: support named parameters
export const findConflictingEndpoints = (endpoints: CompiledEndpoints) => {
  const conflictingEndpoints: string[] = [];
  const registeredRoutes: RegisteredRoutes[] = [];

  for (const [fileName, , extension] of endpoints) {
    const { routeName, method } = convertFileNameToRoute(fileName);
    const conflicts = registeredRoutes.filter(
      route =>
        route.routeName === routeName &&
        (route.method === method || method === "all")
    );

    registeredRoutes.push({
      routeName,
      method,
      fileName: `${fileName}.${extension}`
    });

    if (conflicts.length === 0) continue;

    conflictingEndpoints.push(`${fileName}.${extension}`);

    for (const conflict of conflicts) {
      if (conflictingEndpoints.includes(conflict.fileName)) continue;

      conflictingEndpoints.push(conflict.fileName);
    }
  }

  return conflictingEndpoints;
};
