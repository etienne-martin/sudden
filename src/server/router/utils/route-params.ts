export const mapRouteParams = (
  route: string,
  callback: (match: string) => string
) => {
  return (
    route
      // Split the string by token
      .split(/(\[.*?])/g)
      // Remove empty strings
      .filter(str => str)
      // Detect params and convert them
      .map(chunk => {
        const match = /\[(.*?)]/g.exec(chunk);

        if (match) return callback(match[1]);

        return chunk;
      })
      .join("")
  );
};
