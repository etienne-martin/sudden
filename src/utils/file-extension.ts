export const getFileExtension = (fileName: string) => {
  const ext = /^.+\.([^.]+)$/.exec(fileName);

  return ext === null ? null : ext[1].toLowerCase();
};
