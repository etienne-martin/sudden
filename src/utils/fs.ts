import nodeFs from "fs";
import { promisify } from "util";

export const fs = {
  exists: promisify(nodeFs.exists),
  writeFile: promisify(nodeFs.writeFile),
  mkdir: promisify(nodeFs.mkdir)
};
