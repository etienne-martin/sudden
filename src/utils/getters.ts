import minimist from "minimist";
import path from "path";
import { Task } from "../typings/tasks";

export const getArguments = (nodeProcess: NodeJS.Process) => {
  const { _: args } = minimist(nodeProcess.argv.slice(2));

  return args;
};

export const getOptions = (nodeProcess: NodeJS.Process) => {
  const { _, ...OPTIONS } = minimist(nodeProcess.argv.slice(2)); // eslint-disable-line

  return OPTIONS;
};

export const getOutputDir = (projectDir: string) => {
  return path.resolve(projectDir, ".sudden");
};

export const getSourceDir = (
  projectDir: string,
  task: string | undefined,
  args: any[]
) => {
  return path.resolve(projectDir, (task ? args[1] : args[0]) || "./");
};

export const getTask = (args: any[]): Task => {
  return ["build", "start", "dev"].includes(args[0]) ? args[0] : "dev";
};

export const getPortFromOptions = (options: { [key: string]: any }) => {
  return (
    parseInt((options.port || options.p || "").toString(), 10) || undefined
  );
};

export const getFrameworkDir = (nodeProcess: NodeJS.Process) => {
  return path.resolve(nodeProcess.mainModule!.filename, "../../"); // eslint-disable-line
};
