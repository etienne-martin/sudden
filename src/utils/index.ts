export { noCacheRequire } from "./no-cache-require";
export { getFileExtension } from "./file-extension";
export { rmRf } from "./rm-rf";
export { unescapeUnicode } from "./unicode";
export { detectPackageManager, getPackageJson } from "./package-manager";
export { logger } from "./logger";
export { fs } from "./fs";
export { semVer } from "./semver";
