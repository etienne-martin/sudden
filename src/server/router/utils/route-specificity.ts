import { mapRouteParams } from "./route-params";
import { convertFileNameToRoute, removeMethodFromFileName } from "./index";

export const calculateRouteSpecificity = (fileName: string) => {
  // "all" method is less specific
  // The more params the less specific
  // The more slash the more specific
  // wildcard slugs are less specific (i.e. [slug].js vs post-[postId].js)
  // the more chars (don't count params) the more specific (i.e. a[slug].js vs a-[slug].js)

  const { routeName, method } = convertFileNameToRoute(fileName);
  let specificity = 0;

  // "all" method is less specific
  if (method === "all") specificity--;

  // The more params the less specific
  specificity -= (fileName.match(/(\[.*?])/g) || []).length;

  // The more slash the more specific
  specificity += (routeName.match(/\//g) || []).length;

  let paramsCounter = mapRouteParams(`/${fileName}`, () => "[PARAM]");

  paramsCounter = removeMethodFromFileName(paramsCounter);
  paramsCounter = `${paramsCounter}/`;

  // wildcard slugs are less specific
  specificity -= (paramsCounter.match(/\/\[PARAM]\//g) || []).length;

  // The more chars the more specific
  let charCounter = removeMethodFromFileName(fileName);

  charCounter = mapRouteParams(charCounter, () => "");
  charCounter = charCounter.replace(/\//g, "");

  specificity += charCounter.length;

  return specificity;
};
