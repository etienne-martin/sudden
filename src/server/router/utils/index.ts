import path from "path";

import { mapRouteParams } from "./route-params";
import { calculateRouteSpecificity } from "./route-specificity";
import { getFileExtension, noCacheRequire } from "../../../utils";

type HttpMethod = "get" | "post" | "put" | "delete" | "patch";

interface Route {
  routeName: string;
  method: HttpMethod | "all";
}

type Endpoint = [string, string, number?];
type Endpoints = Endpoint[];
type CompiledEndpoint = [string, () => any, string];
type CompiledEndpoints = CompiledEndpoint[];

const HTTP_METHODS = ["get", "post", "put", "delete", "patch"];

export const convertFileNameToRoute = (fileName: string): Route => {
  const method = getMethodFromFileName(fileName);
  let routeName = removeMethodFromFileName(fileName);

  // Convert "index" to ""
  if (routeName.endsWith("index")) {
    routeName = routeName.substr(0, routeName.lastIndexOf("/index"));
  }

  // Add leading slash
  routeName = `/${routeName}`;

  // Convert square brackets params to express params
  // [userId] to :userId
  routeName = mapRouteParams(routeName, (match: string) => `:${match}`);

  return {
    routeName,
    method: method || "all"
  };
};

export const removeMethodFromFileName = (fileName: string) => {
  const method = getMethodFromFileName(fileName);

  if (method) {
    return fileName.substr(0, fileName.length - `.${method}`.length);
  }

  return fileName;
};

const getMethodFromFileName = (fileName: string): HttpMethod | undefined => {
  const fileExtension = getFileExtension(fileName);

  if (fileExtension && HTTP_METHODS.includes(fileExtension)) {
    return fileExtension as HttpMethod;
  }
};

// TODO: show error for conflicting endpoints
export const getCompiledEndpoints = (
  outputDir: string
): Promise<CompiledEndpoints> =>
  new Promise(async resolve => {
    const endpoints: Endpoints = [];
    const contextRequire = noCacheRequire(
      path.resolve(outputDir, "./endpoints")
    ).default;

    for (const filePath of contextRequire.keys()) {
      const fileExtension = getFileExtension(filePath);

      if (!fileExtension) continue;

      const endpoint = filePath.substr(
        2,
        filePath.length - (fileExtension.length + 1) - 2
      );

      endpoints.push([endpoint, filePath]);
    }

    const sortedEndpoint = sortEndpoints(endpoints);

    resolve(
      sortedEndpoint.map(([fileName, filePath]) => {
        const fileExtension = getFileExtension(filePath);

        return [
          fileName,
          () => contextRequire(filePath),
          fileExtension
        ] as CompiledEndpoint;
      })
    );
  });

const sortEndpoints = (endpoints: Endpoints): Endpoints => {
  // Sort alphabetically
  let sortedEndpoints = endpoints.sort((a, b) => {
    return a[0].localeCompare(b[0]);
  });

  const pointedEndpoints: Endpoints = sortedEndpoints.map(
    ([fileName, filePath]) => [
      fileName,
      filePath,
      calculateRouteSpecificity(fileName)
    ]
  );

  // Sort by specificity
  sortedEndpoints = pointedEndpoints.sort((a, b) => {
    if (a[2] === undefined || b[2] === undefined) return 0; // TS was complaining
    if (a[2] === b[2]) return 0;

    return a[2] > b[2] ? -1 : 1;
  });

  return sortedEndpoints.map(([fileName, filePath]) => [fileName, filePath]);
};
