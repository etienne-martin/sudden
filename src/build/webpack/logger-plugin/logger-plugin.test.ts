import webpack from "webpack";
import { LoggerPlugin } from "./logger-plugin";
import { fs, logger, rmRf } from "../../../utils";

test("should log webpack events to the console", async done => {
  const spy = jest.spyOn(console, "log").mockImplementation();
  const outputPath = `/tmp/logger-plugin`;
  const entryFilePath = `${outputPath}/entry.js`;

  await rmRf(outputPath);
  await fs.mkdir(outputPath);
  await fs.writeFile(entryFilePath, "export default 1;");

  const compiler = webpack({
    mode: "production",
    entry: entryFilePath,
    output: {
      path: outputPath
    },
    resolve: {
      extensions: [".js"]
    },
    plugins: [
      new LoggerPlugin({
        logger: logger
      })
    ]
  });

  compiler.run(async (err, stats) => {
    expect(err).toBe(null);
    expect(stats.hasErrors()).toBe(false);

    // TODO: make sure the console output is correct
    expect(spy).toHaveBeenCalledTimes(2);

    spy.mockRestore();

    done();
  });
});

test("should log failed builds to the console", async done => {
  const consoleInfoSpy = jest.spyOn(console, "info").mockImplementation();
  const consoleLogSpy = jest.spyOn(console, "log").mockImplementation();
  const outputPath = `/tmp/logger-plugin`;
  const entryFilePath = `${outputPath}/entry.js`;

  await rmRf(outputPath);
  await fs.mkdir(outputPath);
  await fs.writeFile(entryFilePath, "import test from 'awdawd';");

  const compiler = webpack({
    mode: "development",
    entry: entryFilePath,
    output: {
      path: outputPath
    },
    resolve: {
      extensions: [".js"]
    },
    module: {
      rules: [
        {
          test: /\.(js)$/,
          use: [
            {
              loader: "babel-loader"
            }
          ]
        }
      ]
    },
    plugins: [
      new LoggerPlugin({
        logger: logger
      })
    ]
  });

  compiler.run(async (err, stats) => {
    expect(err).toBe(null);
    expect(stats.hasErrors()).toBe(true);

    // TODO: make sure the console output contains "failed to compile"
    expect(consoleInfoSpy).toHaveBeenCalledTimes(1);

    consoleInfoSpy.mockRestore();
    consoleLogSpy.mockRestore();

    done();
  });
});
