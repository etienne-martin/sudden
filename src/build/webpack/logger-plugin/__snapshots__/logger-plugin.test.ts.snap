// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should log errors to the console 1`] = `
Array [
  Array [
    "[36m[ wait ] [39m",
    "compiling ...",
  ],
]
`;

exports[`should log errors to the console 2`] = `
Array [
  Array [
    "[37m[ info ] [39m",
    "compiled with errors",
  ],
]
`;

exports[`should log errors to the console 3`] = `Array []`;

exports[`should log errors to the console 4`] = `
Array [
  Array [
    "[31m[ error ][39m",
    [ModuleNotFoundError: Module not found: Error: Can't resolve 'awdawd' in '/private/tmp/logger-plugin'],
  ],
]
`;

exports[`should log errors to the console 5`] = `
Array [
  Array [
    "[36m[ wait ] [39m",
    "creating an optimized production build ...",
  ],
]
`;

exports[`should log errors to the console 6`] = `Array []`;

exports[`should log errors to the console 7`] = `Array []`;

exports[`should log errors to the console 8`] = `
Array [
  Array [
    "[31m[ error ][39m",
    "[31mFailed to compile.[39m",
  ],
  Array [
    "[31m[ error ][39m",
    [ModuleNotFoundError: Module not found: Error: Can't resolve 'awdawd' in '/private/tmp/logger-plugin'],
  ],
]
`;

exports[`should log file changes 1`] = `
Array [
  Array [
    "[36m[ wait ] [39m",
    "compiling ...",
  ],
  Array [
    "[32m[ ready ][39m",
    "compiled successfully",
  ],
  Array [
    "[35m[ event ][39m",
    "file change detected: /private/tmp/logger-plugin/entry.js",
  ],
  Array [
    "[36m[ wait ] [39m",
    "compiling ...",
  ],
  Array [
    "[32m[ ready ][39m",
    "compiled successfully",
  ],
]
`;

exports[`should log file changes 2`] = `Array []`;

exports[`should log file changes 3`] = `Array []`;

exports[`should log file changes 4`] = `Array []`;

exports[`should log warnings to the console 1`] = `
Array [
  Array [
    "[36m[ wait ] [39m",
    "creating an optimized production build ...",
  ],
]
`;

exports[`should log warnings to the console 2`] = `
Array [
  Array [
    "[37m[ info ] [39m",
    "compiled with warnings",
  ],
]
`;

exports[`should log warnings to the console 3`] = `
Array [
  Array [
    "[33m[ warn ][39m",
    [ModuleWarning: Module Warning (from ./node_modules/eslint-loader/index.js):

[4m/private/tmp/logger-plugin/entry.js[24m
  [2m1:16[22m  [31merror[39m  Replace \`¬∑1;\` with \`1;‚èé\`  [2mprettier/prettier[22m

[31m[1m‚úñ 1 problem (1 error, 0 warnings)[22m[39m
[31m[1m[22m[39m[31m[1m  1 error and 0 warnings potentially fixable with the \`--fix\` option.[22m[39m
[31m[1m[22m[39m],
  ],
]
`;

exports[`should log warnings to the console 4`] = `Array []`;

exports[`should log webpack events to the console 1`] = `
Array [
  Array [
    "[36m[ wait ] [39m",
    "creating an optimized production build ...",
  ],
  Array [
    "[32m[ ready ][39m",
    "compiled successfully",
  ],
]
`;

exports[`should log webpack events to the console 2`] = `Array []`;

exports[`should log webpack events to the console 3`] = `Array []`;

exports[`should log webpack events to the console 4`] = `Array []`;
